pkg/virtual/apiexport/authorizer/maximal_permission_policy.go

```diff
@@ -0,0 +1,192 @@
+	"k8s.io/apiserver/pkg/authorization/authorizer"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/clusters"
+
+	apisv1alpha1 "github.com/kcp-dev/kcp/pkg/apis/apis/v1alpha1"
+	"github.com/kcp-dev/kcp/pkg/authorization/delegated"
+	apisinformers "github.com/kcp-dev/kcp/pkg/client/informers/externalversions/apis/v1alpha1"
+	"github.com/kcp-dev/kcp/pkg/indexers"
+	dynamiccontext "github.com/kcp-dev/kcp/pkg/virtual/framework/dynamic/context"
+)
+
+const (
+	VirtualAPIExportMaximumPermissionPolicyAuditPrefix   = "virtual.apiexport.maxpermissionpolicy.authorization.kcp.dev/"
+	VirtualAPIExportMaximumPermissionPolicyAuditDecision = VirtualAPIExportMaximumPermissionPolicyAuditPrefix + "decision"
+	VirtualAPIExportMaximumPermissionPolicyAuditReason   = VirtualAPIExportMaximumPermissionPolicyAuditPrefix + "reason"
+)
+
+const notEnforcedMaximumPermissionPolicyReason = "claimed resource does not enforce maximal permission policy"
+
+type maximalPermissionAuthorizer struct {
+	getAPIExport            func(clusterName, apiExportName string) (*apisv1alpha1.APIExport, error)
+	newDeepSARAuthorizer    func(clusterName logicalcluster.Name) (authorizer.Authorizer, error)
+	getAPIExportsByIdentity func(identityHash string) ([]*apisv1alpha1.APIExport, error)
+}
+
+// NewMaximalPermissionAuthorizer creates an authorizer that checks the maximal permission policy
+// for the requested resource if the resource is a claimed resource in the requested API export.
+// The check is omitted if the requested resource itself is not associated with another API export.
+//
+// If the request is a cluster request the authorizer skips authorization if the request is not for a bounded resource.
+// If the request is a wildcard request this check is skipped because no unique API binding can be determined.
+func NewMaximalPermissionAuthorizer(deepSARClient kubernetes.ClusterInterface, apiExportInformer apisinformers.APIExportInformer, apiBindingInformer apisinformers.APIBindingInformer) authorizer.Authorizer {
+	apiExportLister := apiExportInformer.Lister()
+	apiExportIndexer := apiExportInformer.Informer().GetIndexer()
+	apiBindingIndexer := apiBindingInformer.Informer().GetIndexer()
+	indexers.AddIfNotPresentOrDie(apiBindingIndexer, indexers.ClusterScoped())
+	indexers.AddIfNotPresentOrDie(apiExportIndexer, indexers.ClusterScoped())
+
+	return &maximalPermissionAuthorizer{
+		getAPIExport: func(clusterName, apiExportName string) (*apisv1alpha1.APIExport, error) {
+			return apiExportLister.Get(clusters.ToClusterAwareKey(logicalcluster.New(clusterName), apiExportName))
+		},
+		getAPIExportsByIdentity: func(identityHash string) ([]*apisv1alpha1.APIExport, error) {
+			exports, err := apiExportIndexer.ByIndex(indexers.APIExportByIdentity, identityHash)
+			if err != nil {
+				return nil, err
+			}
+			var results []*apisv1alpha1.APIExport
+			for _, obj := range exports {
+				results = append(results, obj.(*apisv1alpha1.APIExport))
+			}
+			return results, nil
```

Comment:
`indexers.ByIndex` offers go generics support to make this a 1-liner 😄

Reply:
uh, nice! fixing!

Reply:
done, nice!